distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/ServerProcessBuilder.java

```diff
@@ -0,0 +1,188 @@
+
+        return envVars;
+    }
+
+    private List<String> getJvmArgs() {
+        Path esHome = processInfo.workingDir();
+        return List.of(
+            "--module-path",
+            esHome.resolve("lib").toString(),
+            // Special circumstances require some modules (not depended on by the main server module) to be explicitly added:
+            "--add-modules=jdk.net", // needed to reflectively set extended socket options
+            // we control the module path, which may have additional modules not required by server
+            "--add-modules=ALL-MODULE-PATH",
+            "-m",
+            "org.elasticsearch.server/org.elasticsearch.bootstrap.Elasticsearch"
+        );
+    }
+
+    private String getCommand() {
+        Path javaHome = PathUtils.get(processInfo.sysprops().get("java.home"));
+
+        boolean isWindows = processInfo.sysprops().get("os.name").startsWith("Windows");
+        return javaHome.resolve("bin").resolve("java" + (isWindows ? ".exe" : "")).toString();
+    }
+
+    /**
+     * Start a server in a new process.
+     *
+     * @return A running server process that is ready for requests
+     * @throws UserException        If the process failed during bootstrap
+     */
+    public ServerProcess start() throws UserException {
+        return start(ProcessBuilder::start);
+    }
+
+    // package private for testing
+    ServerProcess start(ProcessStarter processStarter) throws UserException {
+        Process jvmProcess = null;
+        ErrorPumpThread errorPump;
+
+        boolean success = false;
+        try {
+            jvmProcess = createProcess(getCommand(), getJvmArgs(), getEnvironment(), processStarter);
+            errorPump = new ErrorPumpThread(terminal.getErrorWriter(), jvmProcess.getErrorStream());
+            errorPump.start();
+            sendArgs(serverArgs, jvmProcess.getOutputStream());
+
+            String errorMsg = errorPump.waitUntilReady();
+            if (errorMsg != null) {
+                // something bad happened, wait for the process to exit then rethrow
+                int exitCode = jvmProcess.waitFor();
+                throw new UserException(exitCode, errorMsg);
+            }
+            success = true;
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        } finally {
+            if (success == false && jvmProcess != null && jvmProcess.isAlive()) {
+                jvmProcess.destroyForcibly();
+            }
+        }
+
+        return new ServerProcess(jvmProcess, errorPump);
+    }
+
+    private Process createProcess(String command, List<String> jvmArgs, Map<String, String> environment, ProcessStarter processStarter)
+        throws InterruptedException, IOException {
+
+        var builder = new ProcessBuilder(Stream.concat(Stream.of(command), Stream.concat(jvmOptions.stream(), jvmArgs.stream())).toList());
```

Comment:
Just a very minor, optional nit. If you pass in JVM opts as param createProcess could be static / stateless helping readability.